crates/swc_ecma_lints/src/rules/prefer_regex_literals.rs

```diff
@@ -0,0 +1,269 @@
+use swc_ecma_visit::{noop_visit_type, Visit, VisitWith};
+
+const UNEXPECTED_REG_EXP_MESSAGE: &str =
+    "Use a regular expression literal instead of the 'RegExp' constructor";
+const UNEXPECTED_REDUNDANT_REG_EXP_MESSAGE: &str =
+    "Regular expression literal is unnecessarily wrapped within a 'RegExp' constructor";
+const UNEXPECTED_REDUNDANT_REG_EXP_WITH_FLAGS_MESSAGE: &str =
+    "Use regular expression literal with flags instead of the 'RegExp' constructor";
+
+const MAX_VALID_ARGS_COUNT: usize = 2;
+
+const GLOBAL_THIS: &str = "globalThis";
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct PreferRegexLiteralsConfig {
+    #[serde(rename = "disallowRedundantWrapping")]
+    disallow_redundant_wrapping: Option<bool>,
+}
+
+pub fn prefer_regex_literals(
+    program: &Program,
+    config: &RuleConfig<PreferRegexLiteralsConfig>,
+    top_level_ctxt: SyntaxContext,
+    es_version: EsVersion,
+) -> Option<Box<dyn Rule>> {
+    let rule_reaction = config.get_rule_reaction();
+
+    let disallow_redundant_wrapping = config
+        .get_rule_config()
+        .disallow_redundant_wrapping
+        .unwrap_or(false);
+
+    let top_level_declared_vars: AHashSet<Id> = collect_decls_with_ctxt(program, top_level_ctxt);
+
+    match rule_reaction {
+        LintRuleReaction::Off => None,
+        _ => Some(visitor_rule(PreferRegexLiterals::new(
+            *rule_reaction,
+            disallow_redundant_wrapping,
+            top_level_declared_vars,
+            top_level_ctxt,
+            es_version,
+        ))),
+    }
+}
+#[derive(Debug)]
+enum ArgValueType {
+    Str,
+    RegExp,
+    Another,
+}
+
+#[derive(Debug, Default)]
+struct PreferRegexLiterals {
+    expected_reaction: LintRuleReaction,
+    disallow_redundant_wrapping: bool,
+    top_level_ctxt: SyntaxContext,
+    top_level_declared_vars: AHashSet<Id>,
+    allow_global_this: bool,
+    call_span: Option<Span>,
+    first_arg: Option<ArgValueType>,
+    second_arg: Option<ArgValueType>,
+}
+
+impl PreferRegexLiterals {
+    fn new(
+        expected_reaction: LintRuleReaction,
+        disallow_redundant_wrapping: bool,
+        top_level_declared_vars: AHashSet<Id>,
+        top_level_ctxt: SyntaxContext,
+        es_version: EsVersion,
+    ) -> Self {
+        Self {
+            expected_reaction,
+            disallow_redundant_wrapping,
+            top_level_ctxt,
+            top_level_declared_vars,
+            allow_global_this: es_version < EsVersion::Es2020,
+            call_span: None,
+            first_arg: None,
+            second_arg: None,
+        }
+    }
+
+    fn reset_state(&mut self) {
+        self.call_span = None;
+        self.first_arg = None;
+        self.second_arg = None;
+    }
+
+    fn set_state(&mut self, call_span: Span, args: &[ExprOrSpread]) {
+        self.call_span = Some(call_span);
+
+        if let Some(ExprOrSpread { expr, .. }) = args.get(0) {
+            let value = match expr.as_ref() {
```

Comment:
I think this logic can be extracted as a function?
Seems like same logic is used twice.